design-proposals/instancetype.kubevirt.io/support-declarative-vm-management.md

```diff
@@ -0,0 +1,277 @@
+
+## `spec.{instancetype,preference}.InferFromVolume`
+
+The use of inferFromVolume will no longer result in the Matcher being mutated within the spec.
+
+All runtime derived details will now be populated within the status of the VirtualMachine.
+
+As with controllerRevisionRef above these details will however be written into the spec during a snapshot to ensure we accurately restore the VirtualMachine later.
+
+## `status.{instancetypeRef,preferenceRef}`
+
+New structs will be introduced under the status of a VirtualMachine to capture the details of each resource.
+
+These details include the `Name,` `Kind` and `ControllerRef` for each resource:
+
+```go
+type ControllerRevisionRef struct {
+	// Name of the ControllerRevision
+	Name string `json:"name"`
+}
+
+type InstancetypeRef struct {
+	// Name is the name of the instance type
+	Name string `json:"name"`
+
+	// Kind specifies the kind of instancetype resource referenced
+	Kind string `json:"kind"`
+
+	// ControllerRef specifies the ControllerRevision storing a copy of the object captured
+// when it is first seen by the VirtualMachine controller
+	ControllerRevisionRef ControllerRevisionRef `json:"controllerRevisionRef"`
+}
+
+type PreferenceRef struct {
+	// Name is the name of the preference
+	Name string `json:"name"`
+
+	// Kind specifies the kind of preference resource referenced
+	Kind string `json:"kind"`
+
+	// ControllerRef specifies the ControllerRevision storing a copy of the object captured
+// when it is first seen by the VirtualMachine controller
+	ControllerRevisionRef ControllerRevisionRef `json:"controllerRevisionRef"`
+}
+
+type VirtualMachineStatus struct {
+[..]
+	// InstancetypeRef captures the state of any referenced instance type from the VirtualMachine
+	//+nullable
+	//+optional
+	InstancetypeRef *InstancetypeRef `json:"instancetypeRef,omitempty"`
+
+	// PreferenceRef captures the state of any referenced preference from the VirtualMachine
+	//+nullable
+	//+optional
+	PreferenceRef *PreferenceRef `json:"instancetypeRef,omitempty"
+}
+```
+
+Storing the `Name` and `Kind` is useful in a few ways. Namely when inferring the details of the resource *and* also checking if a user has requested to switch to a new resource by patching the core spec of the VirtualMachine.
+
+## VirtualMachine subresource API to clear `ControllerRevisionRef`
+
+A user should not be able to modify the status of a VirtualMachine.
+
+As such a new `refresh-{instancetype,preference}` subresource API will be provided to allow users to clear both `spec.{instancetype,preference}.controllerRevsisionRef` and `status.{instancetypeRef,preferenceRef}.controllerRevsisionRef` in order to allow users to move between generations of an instance type or preference object.
+
+A new `virtctl refresh instancetype ${vm}` command should also be created to help users call this subresource API.
```

Comment:
What would be an alternative to `virtctl refresh`? If user wants to not use this subresource API, is it possible?

Reply:
I will also trigger when changing `spec.instancetype`/`spec.`preference` to point to another resource. But when the same resource should be kept I don't see another way.

Reply:
@codingben Yes users can call the subresource API directly but that's not something we typically document in a design proposal.

@0xFelix We wouldn't call the subresource API but similar logic will exist in the VM controller to clear these fields when pointing to a new resource yes.